
{"source":"http://127.0.0.1:<!port!>/uff/1.0/districts/Albany/schools/010/segments/Guardian?sendFlatSpecs=true","destination":"Guardians", "path":"Data"},
{"source":"http://127.0.0.1:<!port!>/uff/1.0/districts/Albany/schools/010/segments/Schedule?sendFlatSpecs=true","destination":"Schedules", "path":"Data"},
{"source":"http://127.0.0.1:<!port!>/uff/1.0/districts/Albany/schools/010/segments/Student_Enrolled?sendFlatSpecs=true","destination":"Students", "path":"Data"},

{"source":"http://127.0.0.1:<!port!>/uff/1.0/districts/Albany/schools/030/segments/Guardian?sendFlatSpecs=true","destination":"Guardians", "path":"Data"},
{"source":"http://127.0.0.1:<!port!>/uff/1.0/districts/Albany/schools/030/segments/Schedule?sendFlatSpecs=true","destination":"Schedules", "path":"Data"},
{"source":"http://127.0.0.1:<!port!>/uff/1.0/districts/Albany/schools/030/segments/Student_Enrolled?sendFlatSpecs=true","destination":"Students", "path":"Data"},

{"source":"http://127.0.0.1:<!port!>/uff/1.0/districts/Albany/schools/035/segments/Guardian?sendFlatSpecs=true","destination":"Guardians", "path":"Data"},
{"source":"http://127.0.0.1:<!port!>/uff/1.0/districts/Albany/schools/035/segments/Schedule?sendFlatSpecs=true","destination":"Schedules", "path":"Data"},
{"source":"http://127.0.0.1:<!port!>/uff/1.0/districts/Albany/schools/035/segments/Student_Enrolled?sendFlatSpecs=true","destination":"Students", "path":"Data"},

{"source":"http://127.0.0.1:<!port!>/uff/1.0/districts/Albany/schools/050/segments/Guardian?sendFlatSpecs=true","destination":"Guardians", "path":"Data"},
{"source":"http://127.0.0.1:<!port!>/uff/1.0/districts/Albany/schools/050/segments/Schedule?sendFlatSpecs=true","destination":"Schedules", "path":"Data"},
{"source":"http://127.0.0.1:<!port!>/uff/1.0/districts/Albany/schools/050/segments/Student_Enrolled?sendFlatSpecs=true","destination":"Students", "path":"Data"},

{"source":"http://127.0.0.1:<!port!>/uff/1.0/districts/Albany/schools/060/segments/Guardian?sendFlatSpecs=true","destination":"Guardians", "path":"Data"},
{"source":"http://127.0.0.1:<!port!>/uff/1.0/districts/Albany/schools/060/segments/Schedule?sendFlatSpecs=true","destination":"Schedules", "path":"Data"},
{"source":"http://127.0.0.1:<!port!>/uff/1.0/districts/Albany/schools/060/segments/Student_Enrolled?sendFlatSpecs=true","destination":"Students", "path":"Data"}