
{"source":"http://127.0.0.1:<!port!>/uff/1.0/districts/Browns%20Valley/schools/010/segments/Guardian?sendFlatSpecs=true","destination":"Guardians", "path":"Data"},
{"source":"http://127.0.0.1:<!port!>/uff/1.0/districts/Browns%20Valley/schools/010/segments/Schedule?sendFlatSpecs=true","destination":"Schedules", "path":"Data"},
{"source":"http://127.0.0.1:<!port!>/uff/1.0/districts/Browns%20Valley/schools/010/segments/Student_Enrolled?sendFlatSpecs=true","destination":"Students", "path":"Data"},
{"source":"http://127.0.0.1:<!port!>/uff/1.0/districts/Browns%20Valley/schools/020/segments/Guardian?sendFlatSpecs=true","destination":"Guardians", "path":"Data"},
{"source":"http://127.0.0.1:<!port!>/uff/1.0/districts/Browns%20Valley/schools/020/segments/Schedule?sendFlatSpecs=true","destination":"Schedules", "path":"Data"},
{"source":"http://127.0.0.1:<!port!>/uff/1.0/districts/Browns%20Valley/schools/020/segments/Student_Enrolled?sendFlatSpecs=true","destination":"Students", "path":"Data"},
{"source":"http://127.0.0.1:<!port!>/uff/1.0/districts/Browns%20Valley/schools/999/segments/Guardian?sendFlatSpecs=true","destination":"Guardians", "path":"Data"},
{"source":"http://127.0.0.1:<!port!>/uff/1.0/districts/Browns%20Valley/schools/999/segments/Schedule?sendFlatSpecs=true","destination":"Schedules", "path":"Data"},
{"source":"http://127.0.0.1:<!port!>/uff/1.0/districts/Browns%20Valley/schools/999/segments/Student_Enrolled?sendFlatSpecs=true","destination":"Students", "path":"Data"}

