
{"source":"http://127.0.0.1:<!port!>/uff/1.0/districts/Paynesville/schools/001/segments/Guardian?sendFlatSpecs=true","destination":"Guardians", "path":"Data"},
{"source":"http://127.0.0.1:<!port!>/uff/1.0/districts/Paynesville/schools/001/segments/Schedule?sendFlatSpecs=true","destination":"Schedules", "path":"Data"},
{"source":"http://127.0.0.1:<!port!>/uff/1.0/districts/Paynesville/schools/001/segments/Student_Enrolled?sendFlatSpecs=true","destination":"Students", "path":"Data"},

{"source":"http://127.0.0.1:<!port!>/uff/1.0/districts/Paynesville/schools/002/segments/Guardian?sendFlatSpecs=true","destination":"Guardians", "path":"Data"},
{"source":"http://127.0.0.1:<!port!>/uff/1.0/districts/Paynesville/schools/002/segments/Schedule?sendFlatSpecs=true","destination":"Schedules", "path":"Data"},
{"source":"http://127.0.0.1:<!port!>/uff/1.0/districts/Paynesville/schools/002/segments/Student_Enrolled?sendFlatSpecs=true","destination":"Students", "path":"Data"},

{"source":"http://127.0.0.1:<!port!>/uff/1.0/districts/Paynesville/schools/020/segments/Guardian?sendFlatSpecs=true","destination":"Guardians", "path":"Data"},
{"source":"http://127.0.0.1:<!port!>/uff/1.0/districts/Paynesville/schools/020/segments/Schedule?sendFlatSpecs=true","destination":"Schedules", "path":"Data"},
{"source":"http://127.0.0.1:<!port!>/uff/1.0/districts/Paynesville/schools/020/segments/Student_Enrolled?sendFlatSpecs=true","destination":"Students", "path":"Data"}

