


{"source":"http://localhost:<!port!>/uff/1.0/districts/0146/schools/0146_010/segments/Guardian?sendFlatSpecs=true?sendFlatSpecs=true","destination":"Guardians", "path":"Data"},
{"source":"http://localhost:<!port!>/uff/1.0/districts/0146/schools/0146_010/segments/Schedule?sendFlatSpecs=true?sendFlatSpecs=true","destination":"Schedules", "path":"Data"},
{"source":"http://localhost:<!port!>/uff/1.0/districts/0146/schools/0146_010/segments/Student_Enrolled?sendFlatSpecs=true?sendFlatSpecs=true","destination":"Students", "path":"Data"},
{"source":"http://localhost:<!port!>/uff/1.0/districts/0146/schools/0146_040/segments/Guardian?sendFlatSpecs=true?sendFlatSpecs=true","destination":"Guardians", "path":"Data"},
{"source":"http://localhost:<!port!>/uff/1.0/districts/0146/schools/0146_040/segments/Schedule?sendFlatSpecs=true?sendFlatSpecs=true","destination":"Schedules", "path":"Data"},
{"source":"http://localhost:<!port!>/uff/1.0/districts/0146/schools/0146_040/segments/Student_Enrolled?sendFlatSpecs=true?sendFlatSpecs=true","destination":"Students", "path":"Data"}


