// dataBuffer/dataBuffer.js
'use strict';
var qtools = require('qtools'),
	qtools = new qtools(module),
	events = require('events'),
	util = require('util'),
	fs=require('fs');

//START OF moduleFunction() ============================================================

var moduleFunction = function(args) {
	events.EventEmitter.call(this);
	this.forceEvent = forceEvent;
	this.args = args;
	this.ping=qtools.ping;

	qtools.validateProperties({
		subject: args,
		targetScope: this, //will add listed items to targetScope
		propList: [
			{
				name:'fileName',
				optional:true
			},
			{
				name:'outputSpec',
				optional:true
			},
			{
				name:'config',
				optional:true
			}
		]
	});

	//note, this is called without parameters for utility storage

	var self = this,
		forceEvent = function(eventName, outData) {
			this.emit(eventName, {
				eventName: eventName,
				data: outData
			});
		};
		
	//LOCAL FUNCTIONS ====================================
	

	//METHODS AND PROPERTIES ====================================
	
	self.alreadyKeptOne=false;
	
	this.dataBuffer='';
	this.dataList=[];
	
	this.hasData=function(){
		return self.dataList.length>0;
	}
	
	this.stash=function(inData, fileWriteCallback){
		self.dataList=self.dataList.concat(inData);
		if (typeof(fileWriteCallback)=='function'){
			fileWriteCallback('', {status:1});
		}
	}
	
	this.getBufferData=function(){
		return this.dataList;
	}
	
	this.takeFlatSpecs=function(flatSpecs){
		self.flatSpecs=flatSpecs;
	}
	this.giveFlatSpecs=function(){
		return self.flatSpecs;
	}

	//INITIALIZATION ====================================

	return this;
};

//END OF moduleFunction() ============================================================

util.inherits(moduleFunction, events.EventEmitter);
module.exports = moduleFunction;
