'use strict';
var qtools = require('qtools');
qtools = new qtools(this);

//source file schemata: Note: These field names are correct according to the new standard
var Address_Contact_FieldList = ["districtCode", "districtType", "studentUniqueIdentifier", "stateStudentIdentifier", "localStudentIdentifier", "staffUniqueIdentifier", "schoolCode", "addressTypeCode", "address1", "address2", "address3", "city", "stateName", "stateCode", "zipCode", "countyCode", "countyName", "contactCode", "contactLegalCode", "contactExtraMailingCode", "contactSequence", "contactFirstName", "contactMiddleName", "contactLastName", "contactFullName", "contactPhone1", "contactPhone2", "contactPhone3", "phone1Type", "phone2Type", "phone3Type", "contactEmail"];
var Course_FieldList = ["districtCode", "districtType", "schoolCode", "courseNumber", "schoolYearBegin", "schoolYearEnd", "courseDesc", "courseAbbreviation", "subjectAreaDepartment", "credit", "gradeLevelMinimum", "gradeLevelMaximum"];
var User_Base_FieldList = ["districtCode", "Filler2", "staffUniqueIdentifier", "schoolYearBegin", "schoolYearEnd", "employeeIdentifier", "localStaffCode", "stateSchoolPlantNumber", "lastName", "middleName", "firstName", "fullName", "jobCode1", "jobCode2", "jobCode3", "jobCode4", "phone", "email", "department1", "department2", "department3", "department4", "Status", "primaryLocaltionFlag", "loginName", "password", "defaultpassword"];
var Lookups_FieldList = ['noFieldDefinitionKnown'];
var Section_FieldList = ["districtCode", "districtType", "schoolCode", "sectionNumber", "courseNumber", "termAbbreviation", "schoolYearBegin", "schoolYearEnd", "gradeLevel", "beginDate", "endDate", "location", "homeroomFlag", "beginningPeriodNumber", "endPeriodNumber", "credit", "teamCode", "track"];
var Section_Staff_FieldList = ["districtCode", "districtType", "schoolCode", "staffUniqueIdentifier", "localStaffCode", "schoolYearBegin", "schoolYearEnd", "sectionNumber", "courseNumber", "gradeLevel", "primaryInstructorFlag", "lastName", "middleName", "firstName"];
var Section_Student_FieldList = ["districtCode", "districtType", "schoolCode", "studentUniqueIdentifier", "stateStudentIdentifier", "localStudentIdentifier", "schoolYearBegin", "schoolYearEnd", "sectionNumber", "courseNumber", "gradeLevelLevel", "entryDate", "withdrawalDate", "entryType", "withdrawalType"];
var Student_Attendance_FieldList = ["districtCode", "districtType", "studentUniqueIdentifier", "stateStudentIdentifier", "localStudentIdentifier", "schoolYearBegin", "schoolYearEnd", "schoolCode", "attendanceCode", "termAbbreviation", "attendanceCategory", "attendanceDate", "attendanceDays", "attendanceTimeUnit", "periodCode", "timeIn", "timeOut", "note"];
var Student_Base_FieldList = ["districtCode", "districtType", "schoolCode", "schoolYearBegin", "schoolYearEnd", "studentUniqueIdentifier", "stateStudentIdentifier", "localStudentIdentifier", "gradeLevelLevel", "graduationYear", "studentActiveInactiveStatus", "lastName", "middleName", "firstName", "suffix", "prefix", "fullName", "preferredName", "lastSchoolAttended", "concurrentEnrollment", "birthDate", "selfGuardianFlag", "gender", "socialSecurityNumber", "birthPlace", "birthState", "birthCountry", "ethnicityCode"];
var Student_Discipline_FieldList = ["districtCode", "districtType", "studentUniqueIdentifier", "stateStudentIdentifier", "localStudentIdentifier", "schoolYearBegin", "schoolYearEnd", "schoolCode", "disciplineTimeUnit", "termAbbreviation", "disciplineCategory", "disciplineCode", "disciplineDate", "dispositionCode", "dispositionBeginDate", "dispositionendDate", "note"];
var Student_Enrollment_FieldList = ["districtCode", "districtType", "studentUniqueIdentifier", "stateStudentIdentifier", "localStudentIdentifier", "schoolCode", "schoolYearBegin", "schoolYearEnd", "gradeLevelLevel", "entryDate", "withdrawalDate", "entryTypeCode", "withdrawalTypeCode", "studentResidentDistrictType", "studentResidentSchoolCode", "studentResidentSchoolCode", "stateAidCategory", "lastlocationOfAttendance", "percentEnrolled", "attendanceType", "membershipDays", "postSecondaryOption", "pseoHighSchoolParticipationHrs", "homeboundServiceIndicator", "specialEducationEvaluationStatus", "specialEducationInstructionalSetting", "lep", "lepBeginDate", "giftedTalentedParticipation", "gender", "ethnicityCode", "birthDate", "homeprimaryLanguage", "primaryDisability", "transportationCategory", "economicIndicator", "migrantIndicator", "studentTitle1Indicator", "homelessStudentFlag", "transportingDistrictCode", "transportingDistrictType", "wardOfStateFlag", "independentStudyFlag", "supplementalEducationServices", "specialEnrollmentCode", "primarySchoolFlag", "specialEducationFlag", "504Flag", "track", "teamCode", "promotionStatus", "programofStudy", "enrollmentStatus", "Filler", "Filler ", "Filler", "Filler", "Filler", "Filler", "advisorIdName", "Filler", "ellServiceLevel", "hispanicLatino", "americanIndianAlaskaNative", "asian", "blackAfricanAmerican", "nativeHawaiianPacificIslander", "white", "firstUsSchoolEntryDate", "newToCountryFlag", "Filler ", "nextYearSchool"];
var Student_gpa_FieldList = ["districtCode", "districtType", "studentUniqueIdentifier", "stateStudentIdentifier", "localStudentIdentifier", "creditsEarned", "schoolYearBegin", "schoolYearEnd", "gpa1", "gpa2", "schoolLevelName", "programofStudy"];
var Student_gradeLevels_FieldList = ["districtCode", "districtType", "studentUniqueIdentifier", "stateStudentIdentifier", "localStudentIdentifier", "courseCode", "termCode", "subjectAreaDepartment", "gradeLevelLevel", "creditsEarned", "creditsAttempted", "gradeLevelMark", "schoolYearBegin", "schoolYearEnd", "schoolCode", "courseDesc", "sectionNumber", "teacherName"];


//Synthetic file schemata
var Student_Enrolled_FieldList = [ "Student_Base.districtCode", "Student_Base.districtType", "Student_Base.schoolCode", "Student_Base.schoolYearBegin", "Student_Base.expr1", "Student_Enrollment.nextYearSchool", "Student_Base.studentUniqueIdentifier", "Student_Base.gender", "Student_Base.stateStudentIdentifier", "Student_Enrollment.gradeLevelLevel", "Student_Base.graduationYear", "Student_Base.studentActiveInactiveStatus", "Student_Base.lastName", "Student_Base.firstName", "Student_Base.middleName", "Address_Contact.address1", "Address_Contact.address2", "Address_Contact.city", "Address_Contact.stateCode", "Address_Contact.zipCode", "Address_Contact.contactPhone1", "Address_Contact.contactPhone2", "Address_Contact.contactPhone3", "Student_Enrollment.studentResidentDistrictCode", "Student_Enrollment.studentResidentSchoolCode", "Student_Base.homeroomTeacher", "Student_Base.counselorIdentifier", "Student_Base.socialSecurityNumber", "Student_Base.ethnicityCode", "Student_Enrollment.homeprimaryLanguage", "Student_Base.primaryLanguage", "Student_Enrollment.stateAidCategory", "Student_Base.lastSchoolAttended", "Student_Base.concurrentEnrollment", "Student_Base.birthDate", "Student_Base.selfGuardianFlag", "Student_Enrollment.wardOfStateFlag", "Student_Enrollment.specialEducationFlag", "Student_Enrollment.specialEducationEvaluationStatus", "Student_Enrollment.specialEducationInstructionalSetting", "Student_Enrollment.primaryDisability", "Student_Enrollment.[504Flag]", "Student_Enrollment.studentTitle1Indicator", "Student_Enrollment.lep", "Student_Enrollment.economicIndicator", "Student_Enrollment.giftedTalentedParticipation", "Student_Enrollment.migrantIndicator", "Student_Enrollment.promotionStatus", "Student_Base.totalYearAbsencesExcused", "Student_Base.totalYearAbsencesUnexcused", "Student_Base.thisYearIss", "Student_Base.thisYearOss", "Student_Base.thisLastYearPromotionStatus", "Student_Base.lastYearAbsencesExplained", "Student_Base.lastYearAbsencesUnexplained", "Student_Base.lastYearIss", "Student_Base.lastYearOss", "Student_Base.gpa", "Student_Base.credits", "Student_Enrollment.programofStudy"];
var Schedule_FieldList = [ 	"Section_Student.districtCode", 	"Section_Student.districtType", 	"Section_Student.schoolCode", 	"Section_Student.studentUniqueIdentifier", 	"Section_Staff.staffUniqueIdentifier", 	"Section_Student.courseNumber", 	"Section_Student.sectionNumber", 	"Section.termAbbreviation", 	"Section_Student.schoolYearBegin", 	"Section.beginningPeriodNumber", 	"Section.endPeriodNumber", 	"Course.courseAbbreviation", 	"Course.subjectAreaDepartment"];
var Guardian_FieldList = [ "Address_Contact.districtCode", "Address_Contact.districtType", "Address_Contact.studentUniqueIdentifier", "Address_Contact.schoolCode", "Address_Contact.guardianType", "Address_Contact.guardianSequence", "Address_Contact.contactLastName", "Address_Contact.contactFirstName", "Address_Contact.contactMiddleName", "Address_Contact.contactFullName", "Address_Contact.address1", "Address_Contact.address2", "Address_Contact.city", "Address_Contact.stateCode", "Address_Contact.zipCode", "Address_Contact.contactPhone1", "Address_Contact.contactPhone2", "Address_Contact.contactPhone3", "Address_Contact.relationship", "Address_Contact.legalGuardian", "Address_Contact.otherGuardian", "Address_Contact.parentStatus", "Address_Contact.emailAddress"];



var processPassThroughUnchanged = require(__dirname + '/lib/all.passThroughUnchanged.js');

var processForStudents = require(__dirname + '/lib/studentPlansInitial.processForStudents.js');
var processForSchedule = require(__dirname + '/lib/studentPlansInitial.processForSchedule.js');
var processForGuardians = require(__dirname + '/lib/studentPlansInitial.processForGuardians.js');
var processPassThroughUnchanged =  require(__dirname + '/lib/all.passThroughUnchanged.js');

module.exports = {


	//NOTE: maps property are "sourceFileFieldName":"targetJsonPropertyName". Empty map, {}, emits entire fieldlist.
	//ALSO: translations are executed *after* maps are set. Their format is: "targetJsonPropertyName": function
	//Translations are 1) the only way to use a source field twice, and
	//2) the only way to *create* a field that does not map to a source field
	//return '<!omitProperty!>'; will remove the property entirely

	//[doc1] - MN SIS Extract Files - Unified_V9(In Progress)
	//[doc2] - Plans4.x Import File Formats
	//[doc3] - DWextractLayout
 
	"Student_Enrolled": {
		componentSchemaList:[
			{
				schemaName: "Student_Base",
				accessInfo: {
					file: {
						fileStandardName: "Student_Base"
					},
					mssql: {
 						query: "SELECT * FROM ext.StudentBase WHERE  schoolCode='<!schoolCode!>';"
 		//				query: "SELECT ext.StudentBase.* FROM (ext.StudentBase LEFT JOIN ext.AddressContact ON ext.StudentBase.studentuniqueidentifier = ext.AddressContact.studentuniqueidentifier) LEFT JOIN ext.StudentEnrollment ON ext.StudentBase.studentuniqueidentifier = ext.StudentEnrollment.studentuniqueidentifier WHERE ( ( ext.StudentEnrollment.entrydate IN ( SELECT Max(entrydate) FROM ext.StudentEnrollment SE WHERE ext.StudentEnrollment.studentuniqueidentifier = SE.studentuniqueidentifier ) ) AND ( ( ext.AddressContact.addresstypecode ) = 'primary1' OR ( ext.AddressContact.addresstypecode ) IS NULL ) AND ( ( ext.StudentEnrollment.enrollmentstatus ) = 'a' ) ) and ext.StudentBase.schoolCode='<!schoolCode!>';"

 					}
				}
			},
			{
				schemaName: "Address_Contact",
				accessInfo: {
					file: {
						fileStandardName: "Address_Contact"
					},
					mssql: {
 						query: "SELECT * FROM ext.AddressContact WHERE  schoolCode='<!schoolCode!>';"
		//				query: "SELECT  ext.AddressContact.* FROM   (ext.StudentBase         LEFT JOIN ext.AddressContact                ON ext.StudentBase.studentuniqueidentifier =                   ext.AddressContact.studentuniqueidentifier)        LEFT JOIN ext.StudentEnrollment               ON ext.StudentBase.studentuniqueidentifier =                  ext.StudentEnrollment.studentuniqueidentifier WHERE  (           (              ext.StudentEnrollment.entrydate IN (                SELECT Max(entrydate)                FROM ext.StudentEnrollment SE                WHERE ext.StudentEnrollment.studentuniqueidentifier = SE.studentuniqueidentifier              )            )          AND ( ( ext.AddressContact.addresstypecode ) = 'primary1'                 OR ( ext.AddressContact.addresstypecode ) IS NULL )          AND ( ( ext.StudentEnrollment.enrollmentstatus ) = 'a' ) )          and ext.StudentBase.schoolCode='<!schoolCode!>';  ext.StudentBase.* FROM   (ext.StudentBase         LEFT JOIN ext.AddressContact                ON ext.StudentBase.studentuniqueidentifier =                   ext.AddressContact.studentuniqueidentifier)        LEFT JOIN ext.StudentEnrollment               ON ext.StudentBase.studentuniqueidentifier =                  ext.StudentEnrollment.studentuniqueidentifier WHERE  (           (              ext.StudentEnrollment.entrydate IN (                SELECT Max(entrydate)                FROM ext.StudentEnrollment SE                WHERE ext.StudentEnrollment.studentuniqueidentifier = SE.studentuniqueidentifier              )            )          AND ( ( ext.AddressContact.addresstypecode ) = 'primary1'                 OR ( ext.AddressContact.addresstypecode ) IS NULL )          AND ( ( ext.StudentEnrollment.enrollmentstatus ) = 'a' ) )          and ext.StudentBase.schoolCode='<!schoolCode!>';"
					}
				}
			},
			{
				schemaName: "Student_Enrollment",
				accessInfo: {
					file: {
						fileStandardName: "Student_Enrollment"
					},
					mssql: {
 						query: "SELECT * FROM ext.StudentEnrollment WHERE  schoolCode='<!schoolCode!>';"
		//				query: "SELECT  ext.StudentEnrollment.* FROM   (ext.StudentBase         LEFT JOIN ext.AddressContact                ON ext.StudentBase.studentuniqueidentifier =                   ext.AddressContact.studentuniqueidentifier)        LEFT JOIN ext.StudentEnrollment               ON ext.StudentBase.studentuniqueidentifier =                  ext.StudentEnrollment.studentuniqueidentifier WHERE  (           (              ext.StudentEnrollment.entrydate IN (                SELECT Max(entrydate)                FROM ext.StudentEnrollment SE                WHERE ext.StudentEnrollment.studentuniqueidentifier = SE.studentuniqueidentifier              )            )          AND ( ( ext.AddressContact.addresstypecode ) = 'primary1'                 OR ( ext.AddressContact.addresstypecode ) IS NULL )          AND ( ( ext.StudentEnrollment.enrollmentstatus ) = 'a' ) )          and ext.StudentBase.schoolCode='<!schoolCode!>';"
					}
				}
			}
		],
		"fieldList": Student_Enrolled_FieldList,
		"assembler": {
			"expressbook": { //captured by dictionary.js, and instance of the named assembler is created with the params which are specific to that assembler
				name: 'pseudoSql', //'pseudoSql', //note: this refers to an assembler in the dataSource module package. Could refer to a fully qualified path to an assembler elsewhere.
				params: {
					finalProcess: processForStudents
				}
			}
		}
	},

	"Schedule": {
		componentSchemaList: [			{
				schemaName: "Section_Student",
				accessInfo: {
					file: {
						fileStandardName: "Section_Student"
					},
					mssql: {
						query: " SELECT * FROM ext.SectionStudent where schoolCode='<!schoolCode!>';"
					}
				}
			},
			{
				schemaName: "Section_Staff",
				accessInfo: {
					file: {
						fileStandardName: "Section_Staff"
					},
					mssql: {
						query: " SELECT * FROM ext.SectionStaff where schoolCode='<!schoolCode!>';"
					}
				}
			},
			{
				schemaName: "Section",
				accessInfo: {
					file: {
						fileStandardName: "Section"
					},
					mssql: {
						query: "SELECT * FROM ext.Section where schoolCode='<!schoolCode!>';"
					}
				}
			},
			{
				schemaName: "Course",
				accessInfo: {
					file: {
						fileStandardName: "Course"
					},
					mssql: {
						query: "SELECT * FROM ext.Course where schoolCode='<!schoolCode!>';"
		//				query: " SELECT ext.Course.* FROM ext.SectionStudent INNER JOIN ext.SectionStaff ON ext.SectionStudent.coursenumber = ext.SectionStaff.coursenumber AND ext.SectionStudent.sectionnumber = ext.SectionStaff.sectionnumber INNER JOIN ext.Section ON ext.SectionStudent.sectionnumber = ext.Section.sectionnumber AND ext.SectionStudent.coursenumber = ext.Section.coursenumber INNER JOIN ext.Course ON ext.SectionStudent.coursenumber = ext.Course.coursenumber where ext.SectionStudent.schoolCode='<!schoolCode!>';"
					}
				}
			}],
		"fieldList": Schedule_FieldList,
		"assembler": {
			"expressbook": { //captured by dictionary.js, and instance of the named assembler is created with the params which are specific to that assembler
				name: 'pseudoSql', //note: this refers to an assembler in the dataSource module package. Could refer to a fully qualified path to an assembler elsewhere.
				params: {
					finalProcess: processForSchedule
				}
			}
		}
	},

	"Guardian": {
		componentSchemaList: [
			{
				schemaName: "Address_Contact",
				accessInfo: {
					file: {
						fileStandardName: "Address_Contact"
					},
					mssql: {
						query: "select * from <!AddressContact!> as baseData where baseData.schoolCode='<!schoolCode!>';"
					}
				}
			}
		],
		"fieldList": Guardian_FieldList,
		"assembler": {
			"expressbook": { //captured by dictionary.js, and instance of the named assembler is created with the params which are specific to that assembler
				name: 'pseudoSql', //note: this refers to an assembler in the dataSource module package. Could refer to a fully qualified path to an assembler elsewhere.
				params: {
					finalProcess: processForGuardians
				}
			}
		}
	},
	

	"Address_Contact":
	{
		componentSchemaList: [
			{
				schemaName: "Address_Contact",
				accessInfo: {
					file: {
						fileStandardName: "Address_Contact"
					},
					mssql: {
						query: "select * from <!AddressContact!> as baseData where baseData.schoolCode='<!schoolCode!>';"
					}
				}
			}
		],
		"fieldList": Address_Contact_FieldList,
		noAssembler:true,
		"translation": {
			"expressbook": {}
		}
	},

	"Course":
	{
		componentSchemaList: [
			{
				schemaName: "Course",
				accessInfo: {
					file: {
						fileStandardName: "Course"
					},
					mssql: {
						query: "select * from <!Course!> as baseData where baseData.schoolCode='<!schoolCode!>';"
					}
				}
			}
		],
		"fieldList": Course_FieldList,
		noAssembler:true,
		"translation": {
			"expressbook": {}
		}
	},

	"User_Base":
	{
		componentSchemaList: [
			{
				schemaName: "User_Base",
				accessInfo: {
					file: {
						fileStandardName: "User_Base"
					},
					mssql: {
						query: "select * from <!UserBase!> as baseData where baseData.schoolCode='<!schoolCode!>';"
					}
				}
			}
		],
		"fieldList": User_Base_FieldList,
		noAssembler:true,
		"translation": {
			"expressbook": {}
		}
	},

	"Lookups":
	{
		componentSchemaList: [
			{
				schemaName: "Lookups",
				accessInfo: {
					file: {
						fileStandardName: "Lookups"
					},
					mssql: {
						query: "select * from <!Lookups!> as baseData where baseData.schoolCode='<!schoolCode!>';"
					}
				}
			}
		],
		"fieldList": Lookups_FieldList,
		noAssembler:true,
		"translation": {
			"expressbook": {}
		}
	},

	"Section":
	{
		componentSchemaList: [
			{
				schemaName: "Section",
				accessInfo: {
					file: {
						fileStandardName: "Section"
					},
					mssql: {
						query: "select * from <!Section!> as baseData where baseData.schoolCode='<!schoolCode!>';"
					}
				}
			}
		],
		"fieldList": Section_FieldList,
		noAssembler:true,

		"translation": {
			"expressbook": {}
		}
	},

	"Section_Staff":
	{
		componentSchemaList: [
			{
				schemaName: "Section_Staff",
				accessInfo: {
					file: {
						fileStandardName: "Section_Staff"
					},
					mssql: {
						query: "select * from <!SectionStaff!> as baseData where baseData.schoolCode='<!schoolCode!>';"
					}
				}
			}
		],
		"fieldList": Section_Staff_FieldList,
		noAssembler:true,
		"translation": {
			"expressbook": {
			}
		}
	},

	"Section_Student":
	{
		componentSchemaList: [
			{
				schemaName: "Section_Student",
				accessInfo: {
					file: {
						fileStandardName: "Section_Student"
					},
					mssql: {
						query: "select * from <!SectionStudent!> as baseData where baseData.schoolCode='<!schoolCode!>';"
					}
				}
			}
		],
		"fieldList": Section_Student_FieldList,
		noAssembler:true,
		"translation": {
			"expressbook": {
			}
		}
	},

	"Student_Attendance":
	{
		componentSchemaList: [
			{
				schemaName: "Student_Attendance",
				accessInfo: {
					file: {
						fileStandardName: "Student_Attendance"
					},
					mssql: {
						query: "select * from <!StudentAttendance!> as baseData where baseData.schoolCode='<!schoolCode!>';"
					}
				}
			}
		],
		"fieldList": Student_Attendance_FieldList,
		noAssembler:true,
		"translation": {
			"expressbook": {}
		}
	},

	"Student_Base":
	{
		componentSchemaList: [
			{
				schemaName: "Student_Base",
				accessInfo: {
					file: {
						fileStandardName: "Student_Base"
					},
					mssql: {
						query: "select * from <!StudentBase!> as baseData where baseData.schoolCode='<!schoolCode!>';"
					}
				}
			}
		],
		"fieldList": Student_Base_FieldList,
		noAssembler:true,
		"translation": {
			"expressbook": {
			}
		}
	},

	"Student_Discipline":
	{
		componentSchemaList: [
			{
				schemaName: "Student_Discipline",
				accessInfo: {
					file: {
						fileStandardName: "Student_Discipline"
					},
					mssql: {
						query: "select * from <!StudentDiscipline!> as baseData where baseData.schoolCode='<!schoolCode!>';"
					}
				}
			}
		],
		"fieldList": Student_Discipline_FieldList,
		noAssembler:true,
		"translation": {
			"expressbook": {}
		}
	},

	"Student_Enrollment":
	{
		componentSchemaList: [
			{
				schemaName: "Student_Enrollment",
				accessInfo: {
					file: {
						fileStandardName: "Student_Enrollment"
					},
					mssql: {
						query: "select * from <!StudentEnrollment!> as baseData where baseData.schoolCode='<!schoolCode!>';"
					}
				}
			}
		],
		"fieldList": Student_Enrollment_FieldList,
		noAssembler:true,
		"translation": {
			"expressbook": {}
		}
	},

	"Student_gpa":
	{
		componentSchemaList: [
			{
				schemaName: "Student_gpa",
				accessInfo: {
					file: {
						fileStandardName: "Student_GPA"
					},
					mssql: {
						tableStandardName: "StudentGpa",
						query: "select gpa.* from <!StudentBase!> as sb left join <!StudentGpa!> as gpa on gpa.localStudentNumber=sb.localStudentIdentifier where sb.schoolCode='<!schoolCode!>';"
					}
				}
			}
		],
		"fieldList": Student_gpa_FieldList,
		noAssembler:true,
		"translation": {
			"expressbook": {}
		}
	},

	"Student_gradeLevels":
	{
		componentSchemaList: [
			{
				schemaName: "Student_gradeLevels",
				accessInfo: {
					file: {
						fileStandardName: "Student_gradeLevels"
					},
					mssql: {
						query: "select * from <!StudentGradeLevels!> as baseData where baseData.schoolCode='<!schoolCode!>';"
					}
				}
			}
		],
		"fieldList": Student_gradeLevels_FieldList,
		noAssembler:true,
		"translation": {
			"expressbook": {}
		}
	}
};









