// assemblers/pseudoSql/pseudoSql.js
'use strict';
var qtools = require('qtools'),
	qtools = new qtools(module),
	events = require('events'),
	util = require('util');

//START OF moduleFunction() ============================================================

var moduleFunction = function(args) {
	events.EventEmitter.call(this);
	this.forceEvent = forceEvent;
	this.args = args;
	this.metaData = {};
	this.addMeta = function(name, data) {
		this.metaData[name] = data;
	}

		qtools.validateProperties({
			subject: args,
			targetScope: this, //will add listed items to targetScope
			propList: [
				{
					name: 'finalProcess',
					optional: false
				}
			]
		});

	var self = this,
		forceEvent = function(eventName, outData) {
			this.emit(eventName, {
				eventName: eventName,
				data: outData
			});
		};


	//LOCAL FUNCTIONS ====================================


	//METHODS AND PROPERTIES ====================================



	//INITIALIZATION ====================================

	var sqlizerGenerator = require('sqlizer');


	this.execute = function(callbackList) {
		var tables = {}
		var list = callbackList,
			outObj;

		for (var i in list) {
			var element = list[i];
			tables[i] = element;
		}

		this.sqlizer = new sqlizerGenerator({
			tables: tables
		});

		outObj = this.finalProcess(this.sqlizer);

		return outObj;
	}
	return this;
};

//END OF moduleFunction() ============================================================

util.inherits(moduleFunction, events.EventEmitter);
module.exports = moduleFunction;
