// serverProfiles/uffFlat1/uffFlat1.js
'use strict';
var qtools = require('qtools'),
	qtools = new qtools(module),
	events = require('events'),
	util = require('util');

//START OF moduleFunction() ============================================================

var moduleFunction = function(args) {
	events.EventEmitter.call(this);
	this.forceEvent = forceEvent;
	this.args = args;
	this.metaData = {};
	this.addMeta = function(name, data) {
		this.metaData[name] = data;
	}

	qtools.validateProperties({
		subject: args || {},
		targetScope: this, //will add listed items to targetScope
		propList: [
			{
				name: 'placeholder',
				optional: true
			}
		]
	});

	var self = this,
		forceEvent = function(eventName, outData) {
			this.emit(eventName, {
				eventName: eventName,
				data: outData
			});
		};


	//LOCAL FUNCTIONS ====================================


	var tableNameMap={
AddressContact:'extract_AddressContact.txt',
Course:'extract_Course.txt.bak',
Section:'extract_Section.txt.bak',
SectionStaff:'extract_SectionStaff.txt.bak',
SectionStudent:'extract_SectionStudent.txt.bak',
StudentAttendance:'extract_StudentAttendance.txt.bak',
StudentBase:'extract_StudentBase.txt.bak',
StudentDiscipline:'extract_StudentDiscipline.txt.bak',
StudentEnrollment:'extract_StudentEnrollment.txt.bak',
StudentGrades:'extract_StudentGrades.txt.bak',
UserBase:'extract_UserBase.txt'
	}


	//METHODS AND PROPERTIES ====================================


	this.mapDistrictAccessName = function(dictionaryName) {
console.log('dictionaryName='+dictionaryName+'\n');


		return "VP_LD_" + dictionaryName;
	}

	this.mapSegmentAccessName = function(dictionaryName) {
console.log('dictionaryName='+dictionaryName+'\n');


		var mapResult=tableNameMap[dictionaryName];
		
		if (typeof(mapResult)!='undefined'){
			return mapResult;
		}
		else{
			return dictionaryName;
		}
	}

// this.mapDistrictAccessName=function(dictionaryName){
// 	return dictionaryName;
// }
// 
// this.mapSegmentAccessName=function(dictionaryName){
// 	return dictionaryName
// }

	//INITIALIZATION ====================================
	
	this.converterName='textToJson';

	return this;
};

//END OF moduleFunction() ============================================================

util.inherits(moduleFunction, events.EventEmitter);
module.exports = moduleFunction;
