// dataAccess/dataAccess.js
'use strict';
var qtools = require('qtools'),
	qtools = new qtools(module),
	events = require('events'),
	util = require('util');

//START OF moduleFunction() ============================================================

var moduleFunction = function(args) {
	events.EventEmitter.call(this);

	qtools.validateProperties({
		subject: args,
		targetScope: this, //will add listed items to targetScope
		propList: [
			{
				name: 'callback',
				optional: false
			},
			{
				name: 'clientProfile',
				optional: false
			},
			{
				name: 'tableName',
				optional: false
			}
		]
	});

	var self = this,
		forceEvent = function(eventName, outData) {
			this.emit(eventName, {
				eventName: eventName,
				data: outData
			});
		}

 
		//PRIVATE FUNCTIONS ====================================



		//METHODS AND PROPERTIES ====================================

	this.forceEvent = forceEvent;

	//INITIALIZATION ====================================

	this.args = args;
	
		var dataSource = {
		textToJson: require('textToJson'), //this doesn't get used until passed to the commandLIneResponder
		dictionary: require('dictionary')
	};

	dataSource.dictionary = new dataSource.dictionary({
		//		dataDefinition: require("./dataDefinitions/" + dictionaryName + ".js"),
		dataDefinition: require("../../dataDefinitions/" + 'uffDefinition' + ".js"),
		target: 'expressbook',
		skipFirstLine: true
	});
	
	
	if (true || self.clientProfile.dataSource.type == 'file') {
		var fileName = self.clientProfile.dataSource.location + self.tableName + '.' + self.clientProfile.dataSource.fileExtension;
	} else {
		qtools.message("we don't do databases yet");
	}

	var controlObj = {
		apiEndpoint: '/data/API/1/Student',
		outboundFinalObjectName: 'StudentPersonal',
		definitionName: 'student'
	};
	
	var sourceData = new dataSource.textToJson(fileName, dataSource.dictionary.get(controlObj.definitionName))

	sourceData.execute();
	sourceData.on('gotData', function() {


		sourceData.mapFieldNames();
		sourceData.processLines();
		sourceData.convert();
		sourceData.assemble();

		args.callback(sourceData.finishedObject);

	});

	return this;
};

//END OF moduleFunction() ============================================================

util.inherits(moduleFunction, events.EventEmitter);
module.exports = moduleFunction;
