// fileWriter/fileWriter.js
'use strict';
var qtools = require('qtools'),
	qtools = new qtools(module),
	events = require('events'),
	util = require('util'),
	fs=require('fs');

//START OF moduleFunction() ============================================================

var moduleFunction = function(args) {
	events.EventEmitter.call(this);
	this.forceEvent = forceEvent;
	this.args = args;
	this.metaData = {};
	this.addMeta = function(name, data) {
		this.metaData[name] = data;
	}

	qtools.validateProperties({
		subject: args,
		targetScope: this, //will add listed items to targetScope
		propList: [
			{
				name: 'fileName',
				optional: false
			}
		]
	});

	var self = this,
		forceEvent = function(eventName, outData) {
			this.emit(eventName, {
				eventName: eventName,
				data: outData
			});
		};


	if (!self.fileWriteCallback) {
		self.fileWriteCallback = function(err) {

			qtools.dump({
				'\n\n===== receiveData.err =====\n': err
			});
		}
	}
	//LOCAL FUNCTIONS ====================================

	//METHODS AND PROPERTIES ====================================

	this.takeItAway=function(data){
	
		fs.writeFile(self.fileName, data, self.fileWriteCallback);
	}

	//INITIALIZATION ====================================

	return this;
};

//END OF moduleFunction() ============================================================

util.inherits(moduleFunction, events.EventEmitter);
module.exports = moduleFunction;
