// transferPacket/transferPacket.js
'use strict';
var qtools = require('qtools'),
	qtools = new qtools(module),
	events = require('events'),
	util = require('util');

//START OF moduleFunction() ============================================================

var moduleFunction = function(args) {
	events.EventEmitter.call(this);
	var self = this,
		forceEvent = function(eventName, outData) {
			this.emit(eventName, {
				eventName: eventName,
				data: outData
			});
		}

		//INITIALIZE OBJECT ====================================

	this.args = args;
	this.metaData = {};
	this.addMeta = function(name, data) {
		this.metaData[name] = data;
	}

	var template = {
		Date: new Date(),
		Status: 1,
		Data: {},
		Error: {},
		Meta: {},
		Message: ''
	};

	var baseObject = {
		Date: new Date(),
		Status: 1
	};

	this.outObject = baseObject;

	//METHODS ====================================

	self.add = function(name, value) {
		if (typeof (template[name]) == 'object') {
			if (typeof (self.outObject[name]) == 'undefined') {
				self.outObject[name] = template[name];
			}
			qtools.extend(self.outObject[name], value);
		} else {
			self.outObject[name] = value;
		}
		return self;
	}

	self.finishedObject = function() {
		return self.outObject;
	}

	self.reset = function() {
		this.outObject = baseObject;
		return self;
	}

	//BUILD RETURN OBJECT ====================================

	this.forceEvent = forceEvent;


	return this;
};

//END OF moduleFunction() ============================================================

util.inherits(moduleFunction, events.EventEmitter);
module.exports = moduleFunction;
