// transferPacket/transferPacket.js
'use strict';
var qtools = require('qtools'),
	qtools = new qtools(module),
	events = require('events'),
	util = require('util');

//START OF moduleFunction() ============================================================

var moduleFunction = function(args) {
	events.EventEmitter.call(this);
	var self = this,
		forceEvent = function(eventName, outData) {
			this.emit(eventName, {
				eventName: eventName,
				data: outData
			});
		}

		//INITIALIZE OBJECT ====================================

	this.args = args;
	this.ping=qtools.ping;

	var template = {
		Date: new Date(),
		Status: 1,
		Error: {},
		Meta: {},
		Message: ''
	};

	var baseObject = {
		Date: new Date(),
		Status: 1
	};

	this.outObject = baseObject;

	//METHODS ====================================

	self.add = function(name, value) {
		if (typeof (value) == 'object') {
			if (typeof (self.outObject[name]) == 'undefined') {
				var valueType=qtools.toType(value);
				self.outObject[name] = (valueType=='array')?[]:{};
			}
			qtools.extend(self.outObject[name], value);
		} else {
			self.outObject[name] = value;
		}
		return self;
	}
	
	self.addMerge=function(inObj){
		for (var i in inObj){
			var element=inObj[i];
			self.add(i, element);
		}
		
	}

	self.finishedObject = function() {
		return self.outObject;
	}

	self.reset = function() {
		this.outObject = baseObject;
		return self;
	}

	//BUILD RETURN OBJECT ====================================

	this.forceEvent = forceEvent;


	return this;
};

//END OF moduleFunction() ============================================================

util.inherits(moduleFunction, events.EventEmitter);
module.exports = moduleFunction;
