// hybridModel/hybridModel.js
'use strict';
var qtools = require('qtools'),
	qtools = new qtools(module),
	events = require('events'),
	util = require('util');

//START OF moduleFunction() ============================================================

var moduleFunction = function(args) {
	events.EventEmitter.call(this);
	this.forceEvent = forceEvent;
	this.args = args;

	qtools.validateProperties({
		subject: args,
		targetScope: this, //will add listed items to targetScope
		propList: [
			{
				name: 'uriPath',
				optional: false
			},
			{
				name: 'clientProfile',
				optional: false
			},
			{
				name: 'apiDefinition',
				optional: false
			}
		]
	});

	var self = this,
		forceEvent = function(eventName, outData) {
			this.emit(eventName, {
				eventName: eventName,
				data: outData
			});
		};


	//LOCAL FUNCTIONS ====================================


	var convertData = function(accessData, mainCallback) {
		mainCallback('', {
			data: accessData,
			meta: "from convertData() stubb"
		});
	}

	var removeUnauthorizedElements = function(entityNameList, dataName) {
		var allowed = self.clientProfile.dataAccess.allowed[dataName],
			disallowed = self.clientProfile.dataAccess.disallowed[dataName],
			outList = [];

		for (var i = 0, len = entityNameList.length; i < len; i++) {
			var element = entityNameList[i];
			if (self.clientProfile.dataAccess.isAllowed(dataName, element.replace(/\.txt/, ''))) {
				outList.push(element);
			}
		}
		return outList;
	}

	//METHODS AND PROPERTIES ====================================


	this.queryInfo = this.apiDefinition.parse(this.uriPath);


	this.getData = function() {
		switch (self.clientProfile.dataSource.type) {
			case 'file':
				var accessEngine = require("fileEngine");
				break;
			case 'mssql':
				qtools.die("add self.clientProfile.dataSource.loginParameters to clientProfile");
				//var accessEngine = require("mssqlEngine");
				break;
		}


		var dictionary = require("dictionary");
		dictionary = new dictionary({
			dataSource:self.clientProfile.dataSource
		});

		var dataConverter = require("textToJson");
		
		dataConverter = new dataConverter({dictionary:dictionary});

		accessEngine = new accessEngine({
			queryInfo: self.queryInfo,
			clientProfile: self.clientProfile,
			removeUnauthorizedElements: removeUnauthorizedElements,
			dataConverter: dataConverter
		});
		

		self.on('getDataComplete', function(accessResult) {
			if (accessResult.status > 0) {

				qtools.mergeMetaData(accessResult.meta);

				convertData(accessResult.data, function(err, conversionResult) {

					if (!err) {
						self.emit('gotData', {
							result: conversionResult.data,
							meta: qtools.mergeMetaData(conversionResult.meta)
						});
					} else {
						self.emit('badData', {
							result: err.data,
							meta: qtools.mergeMetaData(conversionResult.meta)
						});

					}

				});


			} else {
				self.emit('badData', {
					result: accessResult.data,
					meta: qtools.getMetaData()
				});
			}
			
		});

		self.apiDefinition.checkQuery(function(err, result) {
			if (err) {
				var error = {
					status: -1,
					data: err.data,
					meta: qtools.mergeMetaData(err.meta)
				};

				self.emit('getDataComplete', error);
				return false;

			} else {

				accessEngine
					.executeQuery(function(result) {
					self.emit('getDataComplete', result);
				});

			}
		});
	};

	//INITIALIZATION ====================================

	qtools.addMetaData('queryInfo', this.queryInfo);

	return this;
};

//END OF moduleFunction() ============================================================

util.inherits(moduleFunction, events.EventEmitter);
module.exports = moduleFunction;
